src/mkdocs_gallery/gen_single.py

```diff
@@ -739,6 +739,66 @@ def _reset_cwd_syspath(cwd, path_to_remove):
     os.chdir(cwd)
 
 
+def _parse_code(bcontent, src_file, *, compiler_flags):
+    code_ast = compile(bcontent, src_file, "exec", compiler_flags | ast.PyCF_ONLY_AST, dont_inherit=1)
+    if _needs_async_handling(bcontent, compiler_flags=compiler_flags):
+        code_ast = _apply_async_handling(code_ast, compiler_flags=compiler_flags)
+    return code_ast
+
+
+def _needs_async_handling(bcontent, *, compiler_flags) -> bool:
+    try:
+        compile(bcontent, "<_needs_async_handling()>", "exec", compiler_flags, dont_inherit=1)
+    except SyntaxError as error:
+        # FIXME
+        #  bool(re.match(r"'(await|async for|async with)' outside( async)? function", str(error)))
+        #  asynchronous comprehension outside of an asynchronous function
+        return "async" in str(error)
+    except Exception:
+        return False
+    else:
+        return False
+
+
+def _apply_async_handling(code_ast, *, compiler_flags):
+    async_handling = compile(
+        dedent(
+            """
+                async def __async_wrapper__():
+                    # original AST goes here
+                    return locals()
+                import asyncio as __asyncio__
+                __async_wrapper_locals__ = __asyncio__.run(__async_wrapper__())
+                __async_wrapper_result__ = __async_wrapper_locals__.pop("__async_wrapper_result__", None)
+                globals().update(__async_wrapper_locals__)
+                __async_wrapper_result__
+                """
+        ),
+        "<_apply_async_handling()>",
+        "exec",
+        compiler_flags | ast.PyCF_ONLY_AST,
+        dont_inherit=1,
+    )
+
+    *original_body, last_node = code_ast.body
+    if isinstance(last_node, ast.Expr):
+        # FIXME
+        #  ast.Assign(targets=[ast.Name(id="__async_wrapper_result__", ctx=ast.Store())], value=last_node)
+        last_node = compile(
+            f"__async_wrapper_result__ = ({ast.unparse(last_node)})",
+            "<_apply_async_handling()>",
+            "exec",
+            compiler_flags | ast.PyCF_ONLY_AST,
+            dont_inherit=1,
+        ).body[0]
```

Comment:
For some reason, `ast.Assign`ing the `ast.Expr`ession here does not work. With the code that I nicked from

https://github.com/smarie/mkdocs-gallery/blob/cc5c4ad508c03df38c3f58c2c0b8a956d458b900/src/mkdocs_gallery/gen_single.py#L654-L655

I always end up with something like

```pycon
>>> print(ast.unparse(ast.fix_missing_locations(foo)))
__async_wrapper_result__ = 
f'I waited for {stop - start} seconds!'
```

Note the missing parentheses that would allow this as multiline statement. I need to look into why this is happening and fix it. However, for the first iteration, just unparsing the last node, create the assignment as code, and parse it again, works well enough.

Reply:
Fixed in 031bacdddaeb0a3a85c6f4fa8f089fe5f35a6cf7.